lib/smart_proxy_remote_execution_ssh/api.rb

```diff
@@ -1,9 +1,234 @@
+
+        params = MultiJson.load(env["rack.input"].read)
+        key_file = Proxy::RemoteExecution::Ssh.private_key_file
+
+        methods = %w(publickey)
+        methods.unshift('password') if params["ssh_password"]
+
+        ssh_options = { }
+        ssh_options[:port] = params["ssh_port"] if params["ssh_port"]
+        ssh_options[:keys] = [ key_file ] if key_file
+        ssh_options[:password] = params["ssh_password"] if params["ssh_password"]
+        ssh_options[:passphrase] = params[:ssh_key_passphrase] if params[:ssh_key_passphrase]
+        ssh_options[:keys_only] = true
+        ssh_options[:auth_methods] = methods
+        ssh_options[:verify_host_key] = true
+        ssh_options[:number_of_password_prompts] = 1
+
+        socket = nil
+        if env['WEBRICK_SOCKET']
+          socket = env['WEBRICK_SOCKET']
+        elsif env['rack.hijack?']
+          begin
+            env['rack.hijack'].call
+          rescue NotImplementedError
+          end
+          socket = env['rack.hijack_io']
+        end
+        if !socket
+          return [ 501, "Internal error: request hijacking not available" ]
+        end
+
+        ssh_on_socket(socket, params["command"], params["ssh_user"], params["hostname"], ssh_options)
+        101
+      end
+
+      def ssh_on_socket(socket, command, ssh_user, host, ssh_options)
+        started = false
+        err_buf = ""
+        socket = BufferedSocket.build(socket)
+
+        send_start = -> {
+          unless started
+            started = true
+            socket.enqueue("Status: 101\r\n")
+            socket.enqueue("Connection: upgrade\r\n")
+            socket.enqueue("Upgrade: raw\r\n")
+            socket.enqueue("\r\n")
+          end
+        }
+
+        send_error = -> (code, msg) {
+          socket.enqueue("Status: #{code}\r\n")
+          socket.enqueue("Connection: close\r\n")
+          socket.enqueue("\r\n")
+          socket.enqueue(msg)
+        }
+
+        begin
+          Net::SSH.start(host, ssh_user, ssh_options) do |ssh|
+            channel = ssh.open_channel do |ch|
+              ch.exec(command) do |ch, success|
+                raise "could not execute command" unless success
+
+                ssh.listen_to(socket)
+
+                ch.on_process do
+                  if socket.available > 0
+                    ch.send_data(socket.read_available)
+                  end
+                  if socket.closed?
+                    ch.close
+                  end
+                end
+
+                ch.on_data do |ch2, data|
+                  send_start.call
+                  socket.enqueue(data)
+                end
+
+                ch.on_request('exit-status') do |ch, data|
+                  code = data.read_long
+                  send_start.call if code.zero?
+                  err_buf += "Process exited with code #{code}.\r\n"
+                  ch.close
+                end
+
+                channel.on_request('exit-signal') do |ch, data|
```

Comment:
@mvollmer is there a reason why this is using `channel` while the rest is using `ch`?

Reply:
I am not aware of any reason, this probably comes from some example code that I used to get started.